pydantic_ai_slim/pydantic_ai/ag_ui.py

```diff
@@ -0,0 +1,871 @@
+
+    async def _tool_result_event(
+        self,
+        result: ToolReturnPart,
+        prompt_message_id: str,
+    ) -> AsyncGenerator[BaseEvent, None]:
+        """Convert a tool call result to AG-UI events.
+
+        Args:
+            result: The tool call result to process.
+            prompt_message_id: The message ID of the prompt that initiated the tool call.
+
+        Yields:
+            AG-UI Server-Sent Events (SSE).
+        """
+        yield ToolCallResultEvent(
+            message_id=prompt_message_id,
+            type=EventType.TOOL_CALL_RESULT,
+            role=Role.TOOL.value,
+            tool_call_id=result.tool_call_id,
+            content=result.model_response_str(),
+        )
+
+        # Now check for  AG-UI events returned by the tool calls.
+        content: Any = result.content
+        if isinstance(content, BaseEvent):
+            self._logger.debug('ag-ui event: %s', content)
+            yield content
+        elif isinstance(content, (str, bytes)):  # pragma: no branch
+            # Avoid iterable check for strings and bytes.
+            pass
+        elif isinstance(content, Iterable):  # pragma: no branch
+            item: Any
+            for item in content:  # type: ignore[reportUnknownMemberType]
+                if isinstance(item, BaseEvent):  # pragma: no branch
+                    self._logger.debug('ag-ui event: %s', item)
+                    yield item
+
+    async def _agent_stream(
+        self,
+        run: AgentRun[AgentDepsT, Any],
+        history: _History,
+    ) -> AsyncGenerator[BaseEvent, None]:
+        """Run the agent streaming responses using AG-UI protocol events.
+
+        Args:
+            run: The agent run to process.
+            history: The history of messages and tool calls to use for the run.
+
+        Yields:
+            AG-UI Server-Sent Events (SSE).
+        """
+        node: AgentNode[AgentDepsT, Any] | End[FinalResult[Any]]
+        msg: BaseEvent
+        async for node in run:
+            self._logger.debug('processing node=%r', node)
+            if isinstance(node, CallToolsNode):
+                # Handle tool results.
+                async with node.stream(run.ctx) as handle_stream:
+                    async for event in handle_stream:
+                        if isinstance(event, FunctionToolResultEvent) and isinstance(event.result, ToolReturnPart):
+                            async for msg in self._tool_result_event(event.result, history.prompt_message_id):
+                                yield msg
+            elif isinstance(node, ModelRequestNode):
+                # Handle model requests.
+                stream_ctx: _RequestStreamContext = _RequestStreamContext()
+                request_stream: AgentStream[AgentDepsT]
+                async with node.stream(run.ctx) as request_stream:
+                    agent_event: AgentStreamEvent
```

Comment:
This and `request_stream: AgentStream[AgentDepsT]` are unnecessary I think

Reply:
Helps folks understand the intent and can help with auto completion etc.

I've had cases where a declared type hint has helped catch bugs, so instead of playing the guessing game on if it is or isn't needed, which can change over time, I always declare with a type hint.